File: src/motion_types/scripts/rotation.py
#!/usr/bin/env python3
# license removed for brevity
import rospy
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose
import math
import time 


# Initialize global variables
#x = 0
#y = 0
#yaw = 0

def rotate (velocity_publisher , angular_speed_degree , relative_angle_degree , clockwise) :
    velocity_message = Twist()

    angular_speed=math.radians(abs(angular_speed_degree))

    if(clockwise):
        velocity_message.angular.z= -abs(angular_speed)
    else:
        velocity_message.angular.z= abs(angular_speed)

    loop_rate=rospy.Rate(10)
    t0 = rospy.Time.now().to_sec()

    while True : 
        rospy.loginfo("Robot Rotates")
        velocity_publisher.publish(velocity_message)

        t1 = rospy.Time.now().to_sec()
        current_angle_degree = (t1-t0) * angular_speed_degree
        loop_rate.sleep()

        if (current_angle_degree>relative_angle_degree):
            rospy.loginfo("reached")
            break
    velocity_message.angular.z=0
    velocity_publisher.publish(velocity_message)

def poseCallback(pose_message):
    global x, y, yaw
    x = pose_message.x
    y = pose_message.y
    yaw = pose_message.theta

if __name__ == '__main__':        
    rospy.init_node('Robot_rotate', anonymous=True)

    cmd_vel_topic = '/turtle1/cmd_vel'
    velocity_publisher = rospy.Publisher(cmd_vel_topic, Twist, queue_size=10)

    position_topic = "/turtle1/pose"
    pose_subscriber = rospy.Subscriber(position_topic, Pose, poseCallback)
    time.sleep(2)

    rotate(velocity_publisher, 10.0, 90.0 , False)

File: src/motion_types/src/rotation.cpp
#include "ros/ros.h"
#include "geometry_msgs/Twist.h"
#include "turtlesim/Pose.h"
#include <iostream>
#include <cmath>

ros::Publisher velocity_publisher;
turtlesim::Pose turtlesim_pose;



void move(double angular_speed, double relative_angle, bool clockwise) {
    geometry_msgs::Twist vel_msg;

    // Set linear velocity
    if (clockwise)
        vel_msg.angular.z = abs(angular_speed);
    else
        vel_msg.angular.z = -abs(angular_speed);

    // No movement in y, z directions or angular velocity
    vel_msg.linear.x = 0;
    vel_msg.linear.y = 0;
    vel_msg.linear.z = 0;
    vel_msg.angular.x = 0;
    vel_msg.angular.y = 0;

    double current_angle = 0.0;
    double t0 = ros::Time::now().toSec();

    ros::Rate loop_rate(10);  // Control the loop angular_speed

    do {
        velocity_publisher.publish(vel_msg);   // send x = angular_speed 

        double t1 = ros::Time::now().toSec();

        current_angle = angular_speed * (t1 - t0);  // relative_angle = velocity * time
        ros::spinOnce();
        loop_rate.sleep();
    } while (current_angle < relative_angle);

    // Stop the turtle after reaching the desired relative_angle
    vel_msg.angular.z = 0;
    velocity_publisher.publish(vel_msg);
}

void poseCallback(const turtlesim::Pose::ConstPtr &pose_message) {
    turtlesim_pose.x = pose_message->x;
    turtlesim_pose.y = pose_message->y;
    turtlesim_pose.theta = pose_message->theta;
}



int main(int argc, char **argv) {
    // Initiate new ROS node named "move_straight_line"
    ros::init(argc, argv, "move_rotation");
    ros::NodeHandle n;

    //Initialize variables
    double angular_speed, relative_angle;
    bool clockwise;

    // Define the publisher and subscriber
    velocity_publisher = n.advertise<geometry_msgs::Twist>("/turtle1/cmd_vel", 1000);
    ros::Subscriber pose_subscriber = n.subscribe("/turtle1/pose", 10, poseCallback);

    // Test the code
    ROS_INFO("\n\n\n******START TESTING************\n");
    std::cout << "Enter angular_speed: ";
    std::cin >> angular_speed;
    std::cout << "Enter relative_angle: ";
    std::cin >> relative_angle;
    std::cout << "Forward (1 for yes, 0 for no)?: ";
    std::cin >> clockwise;

    move(angular_speed, relative_angle, clockwise);

    return 0;
}
